# 01 - Data Loading and Understanding

This notebook is part of the Blood Donation Prediction project.

# 1. IMPORTING LIBRARIES

In [ ]:
#Importing pandas library to perform many task in dataset.
import pandas as pd

#Importing numpy library used to perform mathematical operation and scientific computing.
import numpy as np

#Importing matplotlib library used to create graphs,chart, and other visualization
import matplotlib.pyplot as plt

#Importing seaborn library used to perform making stastical graphics
import seaborn as sns

#Importing warnings for disable warnings from the code
import warnings
#to ignore the warnings
warnings.filterwarnings('ignore')

#Importing ProfileReport
from ydata_profiling import ProfileReport
# MinMaxScaler for feature scaling to bring values within a specific range
from sklearn.preprocessing import MinMaxScaler

# Importing train_test_split for splitting data into training and testing sets for model evaluation
from sklearn.model_selection import train_test_split

# Importing LogisticRegression
from sklearn.linear_model import LogisticRegression

# Importing DecisionTreeClassifier
from sklearn.tree import DecisionTreeClassifier

# Importing an ensemble model RandomForestClassifier
from sklearn.ensemble import RandomForestClassifier

# Importing an advanced boosting model GradientBoostingClassifier
from sklearn.ensemble import GradientBoostingClassifier

# Importing an efficient and optimized version of gradient- XGBoost (Classifier)
from xgboost import XGBClassifier

# Importing SVC model from sklearn
from sklearn.svm import SVC

# Importing KNeighborsClassifier
from sklearn.neighbors import KNeighborsClassifier

# Importing GridSearchCV for Hyperparameter tuning
from sklearn.model_selection import GridSearchCV

# Importing RandomizedSearchCV for hyperparameter tuning using randomized search
from sklearn.model_selection import RandomizedSearchCV

# Importing Multi-layer Perceptron (MLP) Classifier
from sklearn.neural_network import MLPClassifier

# Importing performance metrics for model evaluation
# accuracy_score,confusion_matrix,classification_report,etc.
from sklearn.metrics import *

# 2. LOADING THE DATASET

In [ ]:
#loading the data
Data=pd.read_csv('Warm_Up_Predict_Blood_Donations_-_Traning_Data.csv')
Data

,Unnamed: 0,Months since Last Donation,Number of Donations,Total Volume Donated (c.c.),Months since First Donation,Made Donation in March 2007
0,619,2,50,12500,98,1
1,664,0,13,3250,28,1
2,441,1,16,4000,35,1
3,160,2,20,5000,45,1
4,358,1,24,6000,77,0
...,...,...,...,...,...,...
571,698,23,1,250,23,0
572,433,16,3,750,86,0
573,360,21,2,500,52,0
574,541,39,1,250,39,0


# 3. BASIC CHECKS

In [ ]:
# Dropping the Unnamed: 0 column
Data.drop('Unnamed: 0',axis=1,inplace=True)

In [ ]:
# Calling the dataset
Data

,Months since Last Donation,Number of Donations,Total Volume Donated (c.c.),Months since First Donation,Made Donation in March 2007
0,2,50,12500,98,1
1,0,13,3250,28,1
2,1,16,4000,35,1
3,2,20,5000,45,1
4,1,24,6000,77,0
...,...,...,...,...,...
571,23,1,250,23,0
572,16,3,750,86,0
573,21,2,500,52,0
574,39,1,250,39,0


In [ ]:
# Extract first 10 row in the dataset
Data.head(10)

,Months since Last Donation,Number of Donations,Total Volume Donated (c.c.),Months since First Donation,Made Donation in March 2007
0,2,50,12500,98,1
1,0,13,3250,28,1
2,1,16,4000,35,1
3,2,20,5000,45,1
4,1,24,6000,77,0
5,4,4,1000,4,0
6,2,7,1750,14,1
7,1,12,3000,35,0
8,5,46,11500,98,1
9,0,3,750,4,0


In [ ]:
# Extract last 10 rows in the dataset
Data.tail(10)

,Months since Last Donation,Number of Donations,Total Volume Donated (c.c.),Months since First Donation,Made Donation in March 2007
566,23,1,250,23,1
567,23,1,250,23,0
568,23,1,250,23,0
569,23,1,250,23,0
570,23,1,250,23,0
571,23,1,250,23,0
572,16,3,750,86,0
573,21,2,500,52,0
574,39,1,250,39,0
575,72,1,250,72,0


In [ ]:
# Checking the number of rows and columns of the data
Data.shape

(576, 5)

In [ ]:
# Checking the basic information of the data
Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 576 entries, 0 to 575
Data columns (total 5 columns):
 #   Column                       Non-Null Count  Dtype
---  ------                       --------------  -----
 0   Months since Last Donation   576 non-null    int64
 1   Number of Donations          576 non-null    int64
 2   Total Volume Donated (c.c.)  576 non-null    int64
 3   Months since First Donation  576 non-null    int64
 4   Made Donation in March 2007  576 non-null    int64
dtypes: int64(5)
memory usage: 22.6 KB


In [ ]:
# Checking the statistical information of the numerical column in the data
Data.describe()

,Months since Last Donation,Number of Donations,Total Volume Donated (c.c.),Months since First Donation,Made Donation in March 2007
count,576.000000,576.000000,576.000000,576.000000,576.000000
mean,9.439236,5.427083,1356.770833,34.050347,0.239583
std,8.175454,5.740010,1435.002556,24.227672,0.427200
min,0.000000,1.000000,250.000000,2.000000,0.000000
25%,2.000000,2.000000,500.000000,16.000000,0.000000
50%,7.000000,4.000000,1000.000000,28.000000,0.000000
75%,14.000000,7.000000,1750.000000,49.250000,0.000000
max,74.000000,50.000000,12500.000000,98.000000,1.000000


In [ ]:
#Extract the columns name
Data.columns

Index(['Months since Last Donation', 'Number of Donations',
       'Total Volume Donated (c.c.)', 'Months since First Donation',
       'Made Donation in March 2007'],
      dtype='object')

In [ ]:
# Checking the index of the data
Data.index

RangeIndex(start=0, stop=576, step=1)

In [ ]:
# Checking the unique values for all the columns in the data
for i in Data:
    print(f"                             {i.title()}                      ")
    print(f"* The number of unique values in {i} column :",Data[i].nunique())
    print(">>",Data[i].unique())
    print('==================================================================================')

                             Months Since Last Donation                      
* The number of unique values in Months since Last Donation column : 29
>> [ 2  0  1  4  5  3 12  6 11  9 10 13 14  8  7 16 23 21 18 22 26 35 74 15
 20 17 25 39 72]
                             Number Of Donations                      
* The number of unique values in Number of Donations column : 29
>> [50 13 16 20 24  4  7 12 46  3 10  6  5 14 15 11  8  9  2 19 17  1 22 18
 38 43 44 21 33]
                             Total Volume Donated (C.C.)                      
* The number of unique values in Total Volume Donated (c.c.) column : 29
>> [12500  3250  4000  5000  6000  1000  1750  3000 11500   750  2500  1500
  1250  3500  3750  2750  2000  2250   500  4750  4250   250  5500  4500
  9500 10750 11000  5250  8250]
                             Months Since First Donation                      
* The number of unique values in Months since First Donation column : 75
>> [98 28 35 45 77  4 14 47 15 11 48 49 16 

In [ ]:
# checking the value counts of all the columns in the data
for i in Data:
    print(f"                             {i.title()}                      ")
    print(Data[i].value_counts())
    print('==================================================================================')

                             Months Since Last Donation                      
Months since Last Donation
2     140
4     115
14     60
11     60
16     48
21     37
23     36
9      17
3      10
1       9
8       5
7       5
12      5
13      4
0       4
6       4
10      3
5       2
18      2
20      1
39      1
25      1
17      1
26      1
15      1
74      1
35      1
22      1
72      1
Name: count, dtype: int64
                             Number Of Donations                      
Number of Donations
1     118
2      85
3      68
5      53
4      49
6      42
7      32
8      27
9      20
11     16
10     11
12      9
16      8
14      8
13      7
17      4
15      3
19      2
24      2
20      2
22      2
43      1
21      1
44      1
50      1
38      1
18      1
46      1
33      1
Name: count, dtype: int64
                             Total Volume Donated (C.C.)                      
Total Volume Donated (c.c.)
250      118
500       85
750       68
1250      53
1000      49


# 4. HANDLING THE DUPLICATES

In [ ]:
# Checking the duplicates values
Data.duplicated().sum()

153

In [ ]:
# Duplicated rows
duplicate_rows=Data[Data.duplicated()]
duplicate_rows

,Months since Last Donation,Number of Donations,Total Volume Donated (c.c.),Months since First Donation,Made Donation in March 2007
16,2,6,1500,15,1
18,2,3,750,4,1
21,2,6,1500,16,1
27,4,10,2500,28,1
36,2,5,1250,16,0
...,...,...,...,...,...
567,23,1,250,23,0
568,23,1,250,23,0
569,23,1,250,23,0
570,23,1,250,23,0


In [ ]:
#Droping the duplicates
Data.drop_duplicates(inplace=True)

In [ ]:
#checking
Data.duplicated().sum()

0

# INSIGHTS
1. The dataset has 576 rows and 5 columns.
3. In the dataset unnnamed: 0 is droped.
2. There are 153 duplicated values  in the dataset.

